### ISSUE TITLE
Improve environment variables support

### ISSUE THREAD

**Issue by brillout** [vikejs/vike#1726]
### Description

- The docs should be improved.
- Create a [new Vike setting `env.public`](https://github.com/vikejs/vike/issues/1724#issuecomment-2202579915).
- How about supporting [`.env.public` files](https://github.com/vikejs/vike/issues/1724#issuecomment-2202551104)?

**Comment by 7iomka** [comment]
<img width="1195" alt="image" src="https://github.com/vikejs/vike/assets/18473137/311a9f21-9501-42a7-847a-3d3d7671ddb7">
Why my `import.meta.env` is undefined?

**Comment by brillout** [comment]
@7iomka It's a Vite design limitation: `import.meta.env` isn't defined in any config file, neither inside Vite's `vite.config.js` file nor inside Vike's `+config.js` files.

Because, in order to determine whether the production or development environment variables should be applied, the value of [Vite's `mode` setting needs to be known](https://vitejs.dev/guide/env-and-mode.html#env-files). Note that [ `mode` can be defined inside `vite.config.js`](https://vitejs.dev/config/shared-options.html#mode).

A possibility would be to stop using Vite's `mode` implemention and have Vike implement its own `mode` feature. (While we could make `import.meta.env` work inside `+config.js` and `vike.config.js` files, I ain't sure we could make it work inside `vite.config.js` files: to be researched.)

**Comment by brillout** [comment]
> Could this be a bug?

Yes, seems like it. Although it could be an issue on Vitest's side. Please re-post on Vike's Github Discussion and publish a reproduction over there, I'll then have a look at it.

**Comment by AgentEnder** [comment]
For server side env vars, is there a reason we can't use the full available environment instead of doing the static replacements? I imagine its different when the server is running etc, but at least for `vike dev` the current approach means that I can't use `zod` to validate / coerce environment variables how I've grown accustomed to.

The pattern I prefer to use looks something like:

```ts
const Env = envSchema.parse(import.meta.env)
```

Where envSchema is a zod object schema

**Comment by brillout** [comment]
@AgentEnder I'll have a look at it.

**Comment by AgentEnder** [comment]
Awesome, for just a tiny bit more context here's a full example:

```typescript
import { z } from 'zod';

const EnvironmentSchema = z.object({
  NODE_ENV: z.string().optional(),
  SQLITE_DB_PATH: z
    .string()
    .optional()
    .describe(
      'Path to the SQLite database file, used for local dev routes to reset the db.'
    ),
  R2_ACCESS_KEY_ID: z.string(),
  R2_SECRET_ACCESS_KEY: z.string(),
  R2_ACCOUNT_ID: z.string(),
});

export type Environment = z.infer<typeof EnvironmentSchema>;
export const Env = EnvironmentSchema.parse(process.env);
```

This is from an express app without vike in the mix, that's bundled with esbuild currently. The project im working on right now is a react-router based SPA admin application, with an express backend. When an admin makes relevant information changes, it kicks off a build of the public vike frontend which uses server side global context to retrieve data from the db and prerender it out. This setup allows db read and write, and api calls, to only ever really be required when an admin is updating things while the public app is a simple static site that can scale out for basically free.

Across the monorepo housing this ive moved the apps to import environment schema fragments and combine them instead of specifying them 100% per app, but not being able to use the object as a whole kinda breaks that approach for the vike frontend, and id eventually like to move the admin site over as well as I do much prefer vike.

**Comment by brillout** [comment]
@AgentEnder How about using `process.env` instead of `import.meta.env`?

**Comment by AgentEnder** [comment]
Would be perfectly fine, its server side anyways. Does that work today in .server.ts files?

**Comment by brillout** [comment]
Yes it should. Let me know if that works for you and I'll update the docs accordingly.

**Comment by nickdichev** [comment]
It should work, I was previously using process.env directly in vike server side code and some places where I haven't migrated to using [@t3-oss/env-core](https://env.t3.gg/docs/core) which wraps standard-schema compliant parsing of env vars -- I've configured that library like this for use with vike:

```javascript
export const env = createEnv({
  server: {
    APP_ENV: z.enum(["local", "test", "production"]).default("local"),
    ...
  },
  client: {
    PUBLIC_ENV__APP_ENV: z.enum(["local", "test", "production"]),
    ...
  },
  runtimeEnv:
    typeof window === "undefined"
      ? process.env
      : {
          PUBLIC_ENV__APP_ENV: import.meta.env.PUBLIC_ENV__APP_ENV,
          ...
        },
  clientPrefix: "PUBLIC_ENV__",
  emptyStringAsUndefined: true
});
```

**Comment by brillout** [comment]
**Edit:** plan moved [here](https://github.com/vikejs/vike/issues/1726#issuecomment-4030273152).

Thanks for sharing, interesting approaches.

Although the drawback of these zod solutions is that they circumvent static replacement, which can be important for code splitting, e.g. `import.meta.env.SSR` should be used over alternatives to eliminate server code from client bundles. See also: https://vike.dev/env#static-replacement.

I've just added env vars defined by `.env` files to `process.env` (https://github.com/vikejs/vike/commit/6117108e5fe3a6bce56964d10c995a363c2e8739) to improve support for both of your zod approaches.

At the moment, these zod approaches are good enough I guess.

More feedback welcome.

**Comment by brillout** [comment]
I just rewrote the docs (https://github.com/vikejs/vike/commit/c9ffd9be52307e54d04e1c92f8f092a2f051c885).

Except of built-in TypeScript support (see [the plan so far](https://github.com/vikejs/vike/issues/1726#issuecomment-3587449815)), I consider this ticket completed.

But feedback welcome.

**Comment by MrNaif2018** [comment]
Hi! Not sure if I should report it here or in a separate issue, but seems related. I was also using [@t3-oss/env-core](https://env.t3.gg/docs/core) for my environment variables parsing like so:

```
import { createEnv } from "@t3-oss/env-core"
import { z } from "zod"

export const env = createEnv({
  client: {
    BITCART_MATOMO_URL: z.string().default(""),
  },

  shared: {
       ...
  },
  clientPrefix: "BITCART_",
  runtimeEnv: import.meta.env,
  emptyStringAsUndefined: true,
})
```

And in vite.config.ts:

```
export default defineConfig({
  envPrefix: "BITCART_",
```

And before 0.4.254 it worked fine, but in 0.4.254 it was added that `import.meta.env` direct invocations get replaced with null because it was confusing why `PUBLIC_` env vars aren't there.
But before the update I had my import.meta.env to an object like:

```
{
  BITCART_MATOMO_URL: 'someurl',
  DEV: false,
  MODE: 'production',
  PROD: true,
  SSR: true
}
```

So t3 env worked correctly. Now I am not sure what is the recommended approach for this, the only way I see it is passing each import.meta.env value to runtimeEnv manually, but I think the previous approach before 0.4.254 worked better

Thank you for your time!

**Comment by brillout** [comment]
Indeed, the bare expression `import.meta.env` is now replaced with `null` by design (implemented at c061f58e313eddede7ef3d67956818ff1d1e8645). @MrNaif2018 Assuming your code is server-only, you can use `process.env` instead of `import.meta.env`. But let me know if that doesn't work for you.

**Comment by MrNaif2018** [comment]
I tried using process.env indeed, and it does seem to output it during prerender stage, but those env variables are needed for me on the client (to enable matomo analytics in my case). Instead of using PUBLIC_ENV_ prefix I configured vite to use BITCART_ prefix, so it worked with import.meta.env before
So during prerender it renders correctly (I can see the variables printed correctly), but when checking on client those env vars are not set unfortunately, so I had to rollback for now until I figure out a better solution
Listing all env vars like so might work
```
runtimeEnv:
    typeof window === "undefined"
      ? process.env
      : {
          BITCART_MATOMO_URL: import.meta.env.BITCART_MATOMO_URL,
          ...
        },
```
but then there is a lot of env vars to list instead of just providing import.meta.env

**Comment by brillout** [comment]
@MrNaif2018 Let's discuss there:
- https://github.com/vikejs/vike/discussions/3146

**Comment by brillout** [comment]
### New plan

Improved DX for zero-config setup:
- At build-time, Vike reads `process.env` and infers the types of all env vars
  - Inferred:
    - If env var value is `"true"` or `"false"` => `boolean` in TS
    - Otherwise, constant `string` in TS. (E.g. `DATABASE_PASSWORD=e9wd!1hw1@#231` => value is `'e9wd!1hw1@#231'` in TS.)
  - The infered types are used as fallback if the env var type isn't already defined via `ImportMetaEnv`
    - Vike codegens a type `ImportMetaEnvWithFallback` and saves it to a file `node_modules/vike/.tmp/env.d.ts`
      - How can Vike add `ImportMetaEnvWithFallback` to `ImportMetaEnv`? Not sure if [this trick](https://chatgpt.com/share/69b03ec1-3f98-800d-8875-a08efaa71baf) works. If it isn't possible => maybe deprecate Vite's `client.d.ts` (not ideal).
  - This means the user can read the *actual* env var values via TypeScript
  - All env vars, including those set via `.env` files and the system (e.g. `$ export SOME_ENV=some-val`)
- If user sets `+env` (see below) => add TypeScript warning via `@deprecated` in JSDocs if the type of the env var isn't defined by a permanent `.env` file (e.g. `.env.local` isn't permanent, and system env vars aren't permanent either) => user must validate it via `+env` (see below)

Validation:
- Via new setting `+env`:
  ```ts
  import * as z from "zod";

  const schema = z.object({
    NODE_ENV: z.enum(["development", "production", "test"]),
    API_URL: z.string().url(),
    PORT: z.string().optional(),
  });

  const jsonSchema = z.toJSONSchema(schema);
  // The list of env vars defined via Zod (or any other tool)
  export const props: string[] = Object.keys(jsonSchema.properties ?? {});

  // Runtime validation via Zod (or any other tool)
  export function validate(env: unknown) {
    return schema.parse(env);
  }
  ```

> [!NOTE]
> - For validation, a tool like zod must be used (or [Telefunc's automatic runtime validation](https://telefunc.com/shield#typescript-automatic)). TypeScript + codegen alone doesn't work for env vars set via the system *after build* (e.g. right before starting the server).
> - The validation can (and should) run exclusively on the server-side (at runtime). (To save client-side KBs, and for security.)

Misc:
- Support runtime env vars on the client-side?
- Vike extensions for tools like https://secretspec.dev/ ?
- Setting to determine what env vars are available to client, e.g. `+env.dangerouslyPassToClient: ['SOME_ENV','SOME_PREFIX_*']`.

**Comment by nickdichev** [comment]
Just to add another perspective, I've solved this issue in a pretty roundabout way but I'm also considering switching my solution as well.

Currently, I am using t3-oss/env-core. What I've done is to configure the server side to read from `process.env` and the client side to read from `window.__RUNTIME_ENV__` which I inject with `+bodyHtmlBegin`. I'm doing this since I build my app with Nix and its pretty annoying to have to build and ship a unique build artifact for each environment (staging, prod, etc) I deploy to.

I'm looking at solutions like https://secretspec.dev since my needs have gotten a little more complicated, and I'm needing to fetch config/secrets from different sources -- not just environment variables.

**Comment by brillout** [comment]
>  annoying to have to build and ship a unique build artifact for each environment

Vike can have have built-in support for runtime env vars on the client-side, but I'm a bit reluctant because it breaks static replacement (tree-shaking). Especially between dev and prod, the user might want tree-shaking. Maybe Vike can support it but with opt-out-able warnings.

> solutions like https://secretspec.dev/

That's a really nice tool, would be lovely if Vike can have some kind of zero-config integration with tools like that via Vike extensions 👀

> Just to add another perspective

Thanks for sharing, that's definitely helpful. More feedback welcome.
